sympy/solvers/solveset.py

```diff
@@ -1096,9 +1096,14 @@ def _solve_exponential(lhs, rhs, symbol):
         return
 
     a, b = list(ordered(lhs.args))
-    log_type_equation = expand_log(log(a)) - expand_log(log(-b))
-
-    return log_type_equation
+    a_coeff, a_term = a.as_independent(symbol)
+    b_coeff, b_term = b.as_independent(symbol)
+
+    if all(term.base.is_positive and term.exp.is_real
+       for term in (a_term, b_term)):
+        log_type_equation = expand_log(log(a)) - expand_log(log(-b))
+        return log_type_equation
```

Comment:
Would it be better to return the conditions under which the solution is valid so a ConditionSet can be created (like the one that log-types create)?

Reply:
Hmmm...since this is the `_solve` helper I think it would be better to return the solution here rather than feeding it back to `_solveset` in `_transolve`:

Reply:
> like the one that log-types create

I don't think log-types creates any kind of `ConditionSet` based on the conditions.

> Would it be better to return the conditions under which the solution is valid so a ConditionSet can be created?

What benefit would it have? we have already restricted solutions of not being in proper assumptions.

Reply:
> .since this is the _solve helper I think it would be better to return the solution here rather than feeding it back to _solveset in _transolve

But that would cause `_solveset` to call here itself (similarly for `_solve_log`) causing repetition (not a big issue but still..).

Reply:
> I don't think log-types creates any kind of ConditionSet based on the conditions.

sorry -- this shows how the distinctions between the set types are not yet well demarcated in my brain! My thinking is far from clear when working with sets. Even when it comes to representing the answer, I see that I easily slip into confusing restrictions on the solutions/x and restrictions on the symbols being used to find x.

To give conditions on the symbols used, it seems (now) that Piecewise would be used. So the solution to `a**(x + 1) - b**x = 0`  would be `Piecewise(({-log(a)/(log(a) - log(b))}, And(a>0, b>0, -oo<x<oo, -oo < x + 1 < oo)), (S.NaN, True))`.

Reply:
What's the best way to say that `f(x)` must be real? `Eq(im(f(x)), 0)`? `-oo < f(x) < oo`?

Reply:
After #14979 was merged, `Eq` will probably be best.

Reply:
This is documentation in a private function, `_transolve`, right? It's fine to change if that's what's necessary.

`fuzzy_and` is looking for True and False; vanilla symbols will not reduce to True/False in inequalities so you will just get None. I'm not sure what definitions you had for a, b, f and g in your example, though.

I would only return a ConditionSet as a Piecewise arg if it represented a solvable expression. If you *know* that `foo(x) = 0` has no solution in a given domain -- not even a numeric approximation -- then IMO you shouldn't return `Eq(f(x), 0)` in any form of the answer.

Reply:
>  It's fine to change if that's what's necessary.

Alright

> I'm not sure what definitions you had for a, b, f and g in your example, though.

Reply:
> t's fine to change if that's what's necessary.

Reply:
> It's fine to change if that's what's necessary.

Alright

> I'm not sure what definitions you had for a, b, f and g in your example, though.

I tried: 

```python

>>> a, b, x = symbols('a b x ')
>>> f = x; g = x + 1    # assuming a**x - b**(x + 1)

>>> cond = And(a>0, b>0, Eq(im(g), 0), Eq(im(f), 0)); cond
(a > 0) & (b > 0) & Eq(im(x), 0)
>>> Piecewise(({log(b)/(log(a) - log(b))}, cond), (ConditionSet(x, Eq(a**x - b**(x + 1)), S.Reals), True))
Piecewise(({log(b)/(log(a) - log(b))}, (a > 0) & (b > 0) & Eq(im(x), 0)), (ConditionSet(x, Eq(a**x - b**(x + 1), 0), Reals), True))

>>> cond2 = True if fuzzy_and([a > 0, b > 0, Eq(im(f), 0), Eq(im(g), 0)]) else False; cond2
False
>>> Piecewise(({log(b)/(log(a) - log(b))}, cond2), (ConditionSet(x, Eq(a**x - b**(x + 1)), S.Reals), True))
ConditionSet(x, Eq(a**x - b**(x + 1), 0), Reals)
```

Reply:
Also if we intend  `_solve_exponential` to return exact solution, then why not `_solve_logarithm` do the same?